File: problems/27-segment-tree/002-lazy-propagation-range-sum/java/Solution.java
public class Solution {
    private long[] tree;
    private long[] lazy;
    private int n;

    public Solution(int[] nums) {
        n = nums.length;
        tree = new long[4 * Math.max(1, n)];
        lazy = new long[4 * Math.max(1, n)];
        if (n > 0) {
            build(nums, 1, 0, n - 1);
        }
    }

    private void build(int[] nums, int node, int left, int right) {
        if (left == right) {
            tree[node] = nums[left];
            return;
        }

        int mid = left + (right - left) / 2;
        build(nums, node * 2, left, mid);
        build(nums, node * 2 + 1, mid + 1, right);
        pull(node);
    }

    // Add delta to every value in [queryLeft, queryRight].
    public void rangeAdd(int queryLeft, int queryRight, long delta) {
        if (n == 0 || queryLeft > queryRight) {
            return;
        }
        rangeAdd(1, 0, n - 1, queryLeft, queryRight, delta);
    }

    private void rangeAdd(int node, int left, int right,
                          int queryLeft, int queryRight, long delta) {
        if (queryLeft <= left && right <= queryRight) {
            tree[node] += delta * (right - left + 1L);
            lazy[node] += delta;
            return;
        }

        push(node, left, right);
        int mid = left + (right - left) / 2;

        if (queryLeft <= mid) {
            rangeAdd(node * 2, left, mid, queryLeft, queryRight, delta);
        }
        if (queryRight > mid) {
            rangeAdd(node * 2 + 1, mid + 1, right, queryLeft, queryRight, delta);
        }

        pull(node);
    }

    // Return the sum of values in [queryLeft, queryRight].
    public long rangeSum(int queryLeft, int queryRight) {
        if (n == 0 || queryLeft > queryRight) {
            return 0;
        }
        return rangeSum(1, 0, n - 1, queryLeft, queryRight);
    }

    private long rangeSum(int node, int left, int right,
                          int queryLeft, int queryRight) {
        if (queryLeft <= left && right <= queryRight) {
            return tree[node];
        }

        push(node, left, right);
        int mid = left + (right - left) / 2;
        long sum = 0;

        if (queryLeft <= mid) {
            sum += rangeSum(node * 2, left, mid, queryLeft, queryRight);
        }
        if (queryRight > mid) {
            sum += rangeSum(node * 2 + 1, mid + 1, right, queryLeft, queryRight);
        }

        return sum;
    }

    private void push(int node, int left, int right) {
        if (lazy[node] == 0 || left == right) {
            return;
        }

        int mid = left + (right - left) / 2;
        long delta = lazy[node];

        tree[node * 2] += delta * (mid - left + 1L);
        tree[node * 2 + 1] += delta * (right - mid);
        lazy[node * 2] += delta;
        lazy[node * 2 + 1] += delta;
        lazy[node] = 0;
    }

    private void pull(int node) {
        tree[node] = tree[node * 2] + tree[node * 2 + 1];
    }
}

File: problems/27-segment-tree/002-lazy-propagation-range-sum/python/solution.py
class Solution:
    def __init__(self, nums: list[int]):
        self.n = len(nums)
        self.tree = [0] * (4 * max(1, self.n))
        self.lazy = [0] * (4 * max(1, self.n))
        if self.n:
            self._build(nums, 1, 0, self.n - 1)

    def _build(self, nums: list[int], node: int, left: int, right: int) -> None:
        if left == right:
            self.tree[node] = nums[left]
            return

        mid = left + (right - left) // 2
        self._build(nums, node * 2, left, mid)
        self._build(nums, node * 2 + 1, mid + 1, right)
        self._pull(node)

    def range_add(self, query_left: int, query_right: int, delta: int) -> None:
        if not self.n or query_left > query_right:
            return
        self._range_add(1, 0, self.n - 1, query_left, query_right, delta)

    def _range_add(
        self,
        node: int,
        left: int,
        right: int,
        query_left: int,
        query_right: int,
        delta: int,
    ) -> None:
        if query_left <= left and right <= query_right:
            self.tree[node] += delta * (right - left + 1)
            self.lazy[node] += delta
            return

        self._push(node, left, right)
        mid = left + (right - left) // 2

        if query_left <= mid:
            self._range_add(node * 2, left, mid, query_left, query_right, delta)
        if query_right > mid:
            self._range_add(node * 2 + 1, mid + 1, right, query_left, query_right, delta)

        self._pull(node)

    def range_sum(self, query_left: int, query_right: int) -> int:
        if not self.n or query_left > query_right:
            return 0
        return self._range_sum(1, 0, self.n - 1, query_left, query_right)

    def _range_sum(
        self,
        node: int,
        left: int,
        right: int,
        query_left: int,
        query_right: int,
    ) -> int:
        if query_left <= left and right <= query_right:
            return self.tree[node]

        self._push(node, left, right)
        mid = left + (right - left) // 2
        total = 0

        if query_left <= mid:
            total += self._range_sum(node * 2, left, mid, query_left, query_right)
        if query_right > mid:
            total += self._range_sum(node * 2 + 1, mid + 1, right, query_left, query_right)

        return total

    def _push(self, node: int, left: int, right: int) -> None:
        if self.lazy[node] == 0 or left == right:
            return

        mid = left + (right - left) // 2
        delta = self.lazy[node]

        self.tree[node * 2] += delta * (mid - left + 1)
        self.tree[node * 2 + 1] += delta * (right - mid)
        self.lazy[node * 2] += delta
        self.lazy[node * 2 + 1] += delta
        self.lazy[node] = 0

    def _pull(self, node: int) -> None:
        self.tree[node] = self.tree[node * 2] + self.tree[node * 2 + 1]
